runtime/src/cost_tracker.rs

```diff
@@ -242,6 +289,32 @@ impl CostTracker {
             .saturating_sub(tx_cost.account_data_size);
         self.transaction_count = self.transaction_count.saturating_sub(1);
     }
+
+    /// Apply execution units adjustment to cost-tracker.
+    /// If adjustment causes arithmetic overflow, it bumps costs to their limits to
+    /// prevent adding further transactions into current block.
+    fn adjust_transaction_execution_cost(&mut self, tx_cost: &TransactionCost, adjustment: i64) {
+        for account_key in tx_cost.writable_accounts.iter() {
+            let account_cost = self
+                .cost_by_writable_accounts
+                .entry(*account_key)
+                .or_insert(0);
+            match ArithmeticUtil::checked_add_signed(*account_cost, adjustment) {
```

Comment:
@t-nelson I don't understand why checking for overflow here instead of saturating is useful. If there's a bug and
- there's overflow, then the behavior is effectively the same as with `saturating_add_signed()`, but it hides the bug from the metrics by setting the cost to a plausible value
- there's an underflow, then that bug now causes the block to stop accepting transactions; not sure if this is better or worse than potentially packing the block too full by saturating to 0

Also I think that doing checked math here, and saturating math in the simple add/remove functions is weirdly inconsistent.

Reply:
Checking potentially allows us to do different handling than just clip to u64::MAX. 

imho, stop accepting TXs to current block when overflow detected is safer than reset cost_tracker to zero therefore producing an oversized block. A underpacked block waster some space, but next block will start anew, while a overpacked block can cause long replay time, and potentially forking. 

But I agree the inconsistency of how `checked_*` and `saturating_*` are used in code, maybe should leave that for another time.